### ISSUE TITLE
can't get it to run

### ISSUE THREAD

**Issue by cj** [dojoteef/neomake-autolint#6]
I re-install neomake with `Plug 'neomake/neomake' | Plug 'dojoteef/neomake-autolint'`.  It doesn't seem to be working as it's not autolinting.

```
± nvim --version
NVIM 0.1.7
Build type: RelWithDebInfo
Compilation: /usr/local/Homebrew/Library/Homebrew/shims/super/clang -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes-std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/tmp/neovim-20161202-25906-1xcct63/neovim-0.1.7/build/config -I/tmp/neovim-20161202-25906-1xcct63/neovim-0.1.7/src -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/gettext/include -I/usr/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include -I/tmp/neovim-20161202-25906-1xcct63/neovim-0.1.7/build/src/nvim/auto -I/tmp/neovim-20161202-25906-1xcct63/neovim-0.1.7/build/include
Compiled by cj@MacBook-Pro.tx.rr.com

Optional features included (+) or not (-): +acl   +iconv    -jemalloc +tui
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.1.7/share/nvim"
```

Please let me know if I can give you any more information.

**Comment by dojoteef** [comment]
@cj Hmm... you have to give me more information than that :-D

Here are examples of useful information you can provide:

Did you read the [FAQ](https://github.com/dojoteef/neomake-autolint#neomake-autolint-is-not-working) regarding what to do if [Neomake Autolint](https://github.com/dojoteef/neomake-autolint) is not working?
What type of file are you trying to lint? 
Does running `:Neomake` from the command line work?

**Comment by cj** [comment]
`:Neomake` shows lint just fine.

`:echo has('nvim') || has('job') && has('timers')` gives an output of 1

I have `let g:neomake_autolint_enabled=1`

I set `let g:neomake_autolint_cachedir='~/.vim/cache'` the folder does exist but I get the following error:

```
error detected while processing function neomake#autolint#Startup[12]..neomake#autolint#Setup[2]..neomake#autolint#utils#cachedir:
line   17:
E739: Cannot create directory ~/.vim/cache/.neomake_autolint_cache: no such file or directory
Press ENTER or type command to continue
```

**Comment by blueyed** [comment]
@cj 
Not sure if the issue initially is about the `mkdir` failure, but that would be fixed by https://github.com/dojoteef/neomake-autolint/pull/7.

**Comment by cj** [comment]
Ok, so the full path fixed that error, it didn't like `~/`.  Now with all those things set, it's still not autolinting.  `:Neomake` works just fine.

**Comment by cj** [comment]
![](http://g.recordit.co/LoVeHnuToS.gif)

**Comment by cj** [comment]
@dojoteef as you can see, it only shows the errors when I `:Neomake` and it seems like `neomake-autolint` is closing everything down once I edit a line but not re-linting.  I have to then manually type `:Neomake` for it to run again.

**Comment by blueyed** [comment]
Woah.. don't know if that is going to tell us that it's workig or not - but the loclist with so much empty lines is annoying.. (that's why I have a local plugin to address this in general, and also Neomake wrappers to only open the list with errors, but not warnings and below).

I would not like for neomake-autolint to mess with the error/location list in general, and that's maybe just what it is doing?!

**Comment by blueyed** [comment]
(it is hard to follow animated gifs like this for me in general - when it is not explained what they are supposed to show etc)

**Comment by cj** [comment]
@blueyed It's just trying to show that autolint isn't linting, I have to manually run `:Neomake`, but it does appear that autolink closes the remanding errors out.  I then have to run `:Neomake` again.

**Comment by blueyed** [comment]
It looked to me like it was adding (error) signs though..?!

**Comment by cj** [comment]
@blueyed only once I run `:Neomake` manually.

**Comment by dojoteef** [comment]
@cj Could please try disabling all other plugins and try again. That can help cut down any potential compatibility issues with other plugins that you have. The reason I ask is it seems your gif shows the location list window closing automatically and Neomake Autolint does not automatically open or close any quickfix/location list windows.

Additionally can you add `let g:neomake_verbose =3` to your .vimrc and post the output of `:messages` after trying to use Neomake Autolint on a file. Additionally please post any configuration settings you have for Neomake and Neomake Autolint in your .vimrc.

**Comment by cj** [comment]
@dojoteef 

```
"~/Sync/apps/acd/m/webpack.config.babel.js" 112L, 3050C
Neomake [2.446]: Toggling buffers: 2
Neomake [2.481]: Executing: autocmd neomake_autolint BufWinEnter,TextChanged,TextChangedI <buffer=2> cal
l s:neomake_onchange(2)
Neomake [2.515]: [1] Running makers: [{'ft': 'javascript.jsx', 'errorformat': '%E%f: line %l\, col %c\,
Error - %m,%W%f: line %l\, col %c\, Warning - %m', 'args': ['-f', 'compact', '/Users/cj/.vim/cache/.neom
ake_autolint_cache/webpack.config.babel.js'], 'mapexpr': 'substitute(v:val, "/Users/cj/.vim/cache/.neoma
ke_autolint_cache/webpack.config.babel.js", l:neomake_bufname, "g")', 'exe': 'eslint', 'append_file': 0,
 'remove_invalid_entries': 1, 'buffer_output': 1, 'name': 'javascript.jsx_eslint'}]
Neomake [2.518]: [1] Starting async job: ['eslint', '-f', 'compact', '/Users/cj/.vim/cache/.neomake_auto
lint_cache/webpack.config.babel.js']
Neomake [12.352]: [1.4] stdout: javascript.jsx_eslint: ['', 'Oops! Something went wrong! :(', '']
Neomake [12.353]: [1.4] stdout: javascript.jsx_eslint: ['', 'ESLint couldn''t find a configuration file.
 To set up a configuration file for this project, please run:', '', '    eslint --init', '', 'ESLint loo
ked for configuration files in /Users/cj/.vim/cache/.neomake_autolint_cache and its ancestors.', '', 'If
 you think you already have a configuration file or if you need more help, please stop by the ESLint cha
t room: https://gitter.im/eslint/eslint', '', '']
Neomake [12.355]: [1.4] exit: javascript.jsx_eslint: 1
Neomake [12.355]: [1.4] javascript.jsx_eslint: processing 11 lines of output.
Neomake [12.356]: [1.4] Removing invalid entry:  ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.356]: [1.4] Removing invalid entry: Oops! Something went wrong! :( ({'lnum': 0, 'bufnr': 0,
'col': 0, 'valid': 0, 'vcol': 0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern':
 ''})
Neomake [12.356]: [1.4] Removing invalid entry:  ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry: ESLint couldn't find a configuration file. To set up a c
onfiguration file for this project, please run: ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol': 0
, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry:  ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry:     eslint --init ({'lnum': 0, 'bufnr': 0, 'col': 0, 'va
lid': 0, 'vcol': 0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry:  ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry: ESLint looked for configuration files in /Users/cj/.vim/
cache/.neomake_autolint_cache and its ancestors. ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
0, 'nr': -1, 'type': '', 'maker_name': 'javascript.jsx_eslint', 'pattern': ''})
Neomake [12.357]: [1.4] Removing invalid entry:  ({'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 0, 'vcol':
```

**Comment by cj** [comment]
```
  let g:neomake_warning_sign = {
  \ 'text': '⚠',
  \ 'texthl': 'WarningMsg',
  \ }
  let g:neomake_error_sign = {
  \ 'text': '✘',
  \ 'texthl': 'ErrorMsg',
  \ }

  let g:neomake_autolint_cachedir='/Users/cj/.vim/cache'
  let g:neomake_open_list=2
  let g:neomake_verbose=3
```

**Comment by dojoteef** [comment]
@cj Hmmm... this looks like it may be specifically ESLint related.

```
Neomake [12.352]: [1.4] stdout: javascript.jsx_eslint: ['', 'Oops! Something went wrong! :(', '']
Neomake [12.353]: [1.4] stdout: javascript.jsx_eslint: ['', 'ESLint couldn''t find a configuration file.
```

The way Neomake Autolint works, it needs to save a temporary file into the cache directory that you set up. It then runs the linter on that file. My guess is that ESLint requires a configuration file in the same directory (or directory hierarchy) as the file it's linting.

@blueyed 
I really wanted to avoid having to make modified versions of makers for Neomake Autolint. I figured it was best to keep this plugin as simple as possible. I can think of a number of potential approaches:

1. Configure linters specially to be able to work with the temporary directory if needed. Con: I would really prefer to avoid the complexity this causes.
2. Have people set up the makers they need themselves in their .vimrc. Con: The more setup needed the more likely people open issues and it's annoying for users to have to figure out how to make their linter work especially if they are unfamiliar with vimscript (even though `:help neomake-configuration` has a good primer)
3. Have Neomake ensure that any additional command line parameters that linters need are set up. For example if a configuration file needs to be specified, make sure the linter has that option set by default automatically even though it is not needed in most cases. Con: Might not work for all cases and it's not clear what project should be responsible for ensuring that happens.

Thoughts?

**Comment by cj** [comment]
@dojoteef the eslint config is always at the root of the project you are working on.  Neomake doesn't have an issue finding the config when running `:Neomake` manually.  `autolint` must be setting/changing that when it runs which is the reason it's not running.....

**Comment by dojoteef** [comment]
@cj As I mentioned the temporary file that Neomake Autolint is writing and
is running the linter on is in your cache directory. For that reason ESLint
cannot find it since the file it is linting is not in your project
directory.

The approach to that problem is what I was discussing with @blueyed.

**Comment by cj** [comment]
@dojoteef I get you.  Sorry about that, long day/night.

**Comment by dojoteef** [comment]
@cj No problem. I understand.

**Comment by dojoteef** [comment]
@cj You could temporarily modify your Neomake maker to point to the `.eslintrc` until I figure out what the best solution to this problem is.

Here is the [current maker definition for eslint](https://github.com/neomake/neomake/blob/master/autoload/neomake/makers/ft/javascript.vim#L32-L38). Simply add the command line argument to point to the eslint configuration you want. See [Configuration ESLint](http://eslint.org/docs/user-guide/configuring) and [ESLint Command Line interface](http://eslint.org/docs/user-guide/command-line-interface).

**Comment by dojoteef** [comment]
@blueyed Any thoughts on [my comment](https://github.com/dojoteef/neomake-autolint/issues/6#issuecomment-265346090)?

**Comment by blueyed** [comment]
@dojoteef 
What about creating the temporary file next to existing one, potentially as a hidden file (prefixed with `.`)?!

And btw: https://github.com/neomake/neomake/pull/867 is about to add support for a callback on makers itself (https://github.com/neomake/neomake/pull/867/files#diff-58e65732fcc7ab450ae138122e6ee651).

There is some discussion at https://github.com/tweekmonster/deoplete-clang2/commit/872ff514168a719d33b8e588bc64962fa75100dd#commitcomment-20225451 about this and using some User autocommand additionally/instead.
I think the latter would help you a lot, where you could change the maker just before it gets executed.

**Comment by dojoteef** [comment]
@cj How are you fairing? Were you able to at least get it working temporarily?

I'm still not sure what the best approach is for the general case and I am hesitant to put a specific fix in for [ESlint] since it potentially requires linter specific overrides for linters that [Neomake] supports. I am totally willing to document how to setup [ESLint] in the documentation as a stopgap until I can come up with an extensible solution. If you have any suggestions I'm willing to listen.

@blueyed I dislike the idea of adding anything to the user's current working directory for fear that they may accidentally add that file into their source code repositories. This could theoretically happen if the linter takes a while to run and they switch to a shell prompt and do a `git add .` because by default git will not ignore the file. I know the likelihood of that situation is low, but it does exist.

One solution is to make use of the [--stdin](http://eslint.org/docs/user-guide/command-line-interface#options) feature of [ESLint]. If [Neomake] supported that by default rather than directly linting the file that would make this "just work" in [Neomake-Autolint].

[ESlint]: http://eslint.org
[Neomake]: https://github.com/neomake/neomake
[Neomake-Autolint]: https://github.com/dojoteef/neomake-autolint

**Comment by blueyed** [comment]
@dojoteef 
> I know the likelihood of that situation is low, but it does exist.

It's not that unlikely, and a fair point against this.

**Comment by cj** [comment]
@dojoteef still no luck.  I just get this now `Neomake [1.139]: Skipping User autocmd NeomakeFinished: no hooks.`.  I can run `:Neomake` manually just fine however.

**Comment by dojoteef** [comment]
@cj Can you provide your updated maker configuration that references your `.eslintrc`?

**Comment by blueyed** [comment]
Also the output from `:NeomakeInfo` might be helpful.

@dojoteef 
I could imagine for neomake-autolint to hook into it somehow.  What do you think?

**Comment by cj** [comment]
@dojoteef @blueyed 

```
#### Neomake debug information
Async support: 1
Current filetype: javascript.jsx

##### Enabled makers
For the current filetype (with :Neomake):
 - eslint
   errorformat: '%E%f: line %l\, col %c\, Error - %m,%W%f: line %l\, col %c\, Warning - %m'
   args: ['-f', 'compact']
   get_argv: function('11')
   exe: 'eslint'
NOTE: you can define g:neomake_javascript.jsx_enabled_makers to configure it (or b:neomake_javascript.jsx_
enabled_makers).

For the project (with :Neomake!): None.
NOTE: you can define g:neomake_enabled_makers to configure it.

Default maker settings:
 - remove_invalid_entries: 0
 - buffer_output: 1

##### Settings
g:neomake_autolint_enabled = 1
g:neomake_autolint_cachedir = '/Users/cj/.vim/cache'
g:neomake_verbose = 3
g:neomake_place_signs = 1
g:neomake_warning_sign = {'texthl': 'WarningMsg', 'text': '⚠'}

shell: /usr/local/bin/zsh
shellcmdflag: -c
Windows: 0
```

The issue is, it's not finding the `.eslintrc.json` config in the root of the project, where I do `vim .` from.  If I add `.eslintrc.json` to my home dir autolint works.

**Comment by dojoteef** [comment]
@cj I don't see an overridden maker in your settings that references the `.eslintrc` file. You may have missed [my suggestion](https://github.com/dojoteef/neomake-autolint/issues/6#issuecomment-268365813) for how to fix this.

Try adding something like this to your `.vimrc`:
```vim
  " Point to the correct .eslintrc file for the buffer.
  function! s:ESLintSetup()
    " NOTE: Correct usage of '/' depends on +shellslash option
    let l:eslintrc = join([getcwd(), '.eslintrc'], '/')
    if filereadable(l:eslintrc)
      let b:neomake_javascript_eslint_args =
            \ ['-f', 'compact', '-c', l:eslintrc]
    endif
  endfunction

  autocmd vimrc FileType javascript
        \ autocmd vimrc User NeomakeAutolint call s:ESLintSetup()
```

Additionally if you wanted more complicated behavior in regards to finding the correct `.eslintrc` file for the project you could make use of the vimscript function `findfile()`.

**NOTE:** I do not develop in Javascript and do not have ESLint installed, but I did test that if there is an `eslint` executable and you open a Javascript file in a directory that has an `.eslintrc` that it will correctly setup the maker args. Please let me know the outcome of trying this.

**EDIT:** Fixed a small issue in the `args`

**Comment by blueyed** [comment]
Neomake supports automaking by now itself.. :fireworks:

**Comment by dojoteef** [comment]
The README now clearly states the plugin is deprecated (I haven't actually had the time to update it for a while now anyway, sorry). If you are having any issues please try Neomake's builtin automake functionality.
